mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,1129 @@
+        block::{calculate_hash, validate, BlockHeaderWithHash, BLOCK_HINT, SERIALIZED_BLOCK_SIZE},
+        error::RingBufferError,
+        flush::{FlushOptions, FlushState},
+        serialize::{binary_deserialize, binary_deserialize_owned, binary_serialize},
+    },
+    Key, StorageError,
+};
+
+pub type StorageResult<T> = Result<T, StorageError>;
+
+/// Convenience struct for tracking read and write pointers into the file.
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub(crate) struct FilePointers {
+    write: usize,
+    read_begin: usize,
+    read_end: usize,
+}
+
+/// Imagine there are three pointers:
+/// W - A write pointer
+/// R - A read pointer
+/// D - A delete pointer
+///
+/// With a file buffer similar to below:
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around and cause us to lose data due to
+/// variable sized data being written. If W + data size > R and should write
+/// precede read, the data could become corrupted.
+///
+/// Either this is prevented by scanning the file for the next block and
+/// checking if the write will fit. Or, W + data must not pass both D and R.
+/// - Not passing D helps with guaranteeing all data read was handled.
+/// - Not passing R helps avoid corruption.
+pub(crate) struct RingBuffer {
+    // If write reaches read but we have data to read, this is set.
+    // It allows us handle the edge case of read == write, but data
+    // has not been read yet.
+    can_read_from_wrap_around: bool,
+    // The optional flush threshold to uphold.
+    flush_options: FlushOptions,
+    // For determining if a flush should occur.
+    flush_state: FlushState,
+    // If data has been read this is set, if data is removed to
+    // the point of reaching the read pointer this is unset.
+    // This prevents deletion of data without reading.
+    has_read: bool,
+    // Max size for the file that is required for mmap.
+    max_file_size: usize,
+    // A representation of Mmap with operations built-in.
+    mmap: MmapMut,
+    // A tracker for suppling an ordering to blocks being written.
+    order: usize,
+    // Pointers into the file (read/write).
+    pointers: FilePointers,
+    // A waker for updating any pending batch after an insert.
+    waker: Option<Waker>,
+}
+
+impl RingBuffer {
+    pub(crate) fn new(
+        file_path: &Path,
+        max_file_size: usize,
+        flush_options: &FlushOptions,
+    ) -> StorageResult<Self> {
+        let file = create_file(file_path).map_err(RingBufferError::MmapCreate)?;
+        file.set_len(max_file_size as u64)
+            .map_err(RingBufferError::MmapCreate)?;
+        let mmap = unsafe { MmapMut::map_mut(&file).map_err(RingBufferError::MmapCreate)? };
+        let (file_pointers, order) = find_pointers_and_order_post_crash(&mmap, max_file_size)?;
+
+        Ok(Self {
+            can_read_from_wrap_around: false,
```

Comment:
I think `find_pointers_and_order_post_crash()` should also determine and return `can_read_from_wrap_around` flag as well. It is possible that the buffer was "full" at the time of shutdown/crash. So next time you start the ring buffer, you would think that it is "empty" since read==write.